### PR TITLE
cl/services: share pending queue for payload gossip

### PR DESCRIPTION
**Pull request by yperbasis**
First PR in the split of #22177.

This PR extracts the duplicated pending scheduler used by execution-payload envelopes and payload attestations into an unexported `pendingJobQueue[K, M]`.

There are no intended behavior changes. Each service keeps its existing key, capacity, 100 ms polling interval, 30-second expiry, duplicate handling, full-queue fast path, removal-before-processing order, cancellation behavior, and logs. The existing service tests are mechanically adapted to the shared representation.

TDD was not used because this is a pure refactor; the existing service tests are the safety net.

## Stack

- #23643 — shared-queue refactor (this PR)
- #23644 — queue hardening and bid migration
- #23645 — deferred Gloas data-column sidecars
- #23646 — deferred beacon blocks

The last two PRs are sibling branches and can be reviewed independently after #23644.

### CODE REVIEW

**Review by AskAlexSharov — APPROVED** [review]

File: cl/phase1/network/services/payload_attestation_service.go [hunk 1]
```diff
@@ -207,23 +205,9 @@ func (s *payloadAttestationService) ProcessMessage(ctx context.Context, _ *uint6
...
-		creationTime: time.Now(),
-	}); loaded {
-		s.pendingCount.Add(-1)
-	} else {
-		s.pendingCond.L.Lock()
-		s.pendingCond.Signal()
-		s.pendingCond.L.Unlock()
-	}
+	_ = s.pending.enqueueLazy(msg, func() (pendingPayloadAttestationKey, error) {
+		return pendingPayloadAttestationKeyFor(blockRoot, msg), nil
```
Comment (AskAlexSharov) [inline cl/phase1/network/services/payload_attestation_service.go:209]: `pendingPayloadAttestationKeyFor` drops the `HashSSZ` error (`root, _ := msg.HashSSZ()`), and this key builder returns `nil` for it, so `queuePendingAttestation` discards the result with `_ =`.

`merkle_tree.HashTreeRoot` can fail. On failure `messageRoot` is the zero hash, so every attestation that fails to hash collapses onto the same `pendingPayloadAttestationKey` and `LoadOrStore` silently drops all but the first — the exact case `TestPayloadAttestationServicePendingQueueKeepsDistinctSameValidatorBlock` pins.

The new lazy-key API exists to report this. The envelope side already uses it; the attestation side should too:

```go
_ = s.pending.enqueueLazy(msg, func() (pendingPayloadAttestationKey, error) {
	root, err := msg.HashSSZ()
	if err != nil {
		return pendingPayloadAttestationKey{}, err
	}
	return pendingPayloadAttestationKey{blockRoot: blockRoot, validatorIndex: msg.ValidatorIndex, messageRoot: common.Hash(root)}, nil
})
```
(and log the returned error, as `queuePendingEnvelope` does).

File: cl/phase1/network/services/pending_job_queue.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+		expiry:     expiry,
+		tick:       tick,
+		tryProcess: tryProcess,
+		onExpired:  onExpired,
+		cond:       sync.NewCond(&sync.Mutex{}),
+	}
+}
+
+func (q *pendingJobQueue[K, M]) enqueueLazy(msg M, buildKey func() (K, error)) error {
+	if !q.reserve() {
```
Comment (AskAlexSharov) [inline cl/phase1/network/services/pending_job_queue.go:66]: A full queue returns `nil`, so the caller cannot tell "queued" from "dropped".

`payloadAttestationService.ProcessMessage` returns `%w: %w: block not available` with `ErrAttestationQueued` right after `queuePendingAttestation`, even when the message was rejected by the cap and never stored. Nothing logs or meters the drop either, so a flood that fills the queue is invisible.

Since this is now a single shared place, a sentinel (`errPendingQueueFull`) would let both services log/meter the drop and let the attestation path stop claiming the message was queued.

File: cl/phase1/network/services/pending_job_queue.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+func (q *pendingJobQueue[K, M]) processPending(ctx context.Context) {
+	q.jobs.Range(func(key, value any) bool {
+		k := key.(K)
+		job := value.(*pendingJob[M])
+
+		if time.Since(job.creationTime) > q.expiry {
+			q.remove(k)
+			q.onExpired(k)
```
Comment (AskAlexSharov) [inline cl/phase1/network/services/pending_job_queue.go:151]: `q.onExpired(k)` (and `q.tryProcess` on line 155) are called with no nil guard from the background `loop` goroutine.

Both current call sites pass non-nil, but this type is being introduced as the shared queue for three more services (#23644, #23645, #23646). A migration that leaves `onExpired` nil panics inside a goroutine with no recover, which takes the whole node down instead of failing at construction.

Either check for nil here, or assert non-nil in `newPendingJobQueue`.

File: cl/phase1/network/services/pending_job_queue.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	key, err := buildKey()
+	if err != nil {
+		q.count.Add(-1)
+		return err
+	}
+	q.storeReserved(key, msg)
+	return nil
+}
+
+func (q *pendingJobQueue[K, M]) reserve() bool {
```
Comment (AskAlexSharov) [inline cl/phase1/network/services/pending_job_queue.go:79]: `reserve()` increments first and compares after, so `count` can transiently exceed `capacity` by the number of concurrent enqueuers, and enqueues get rejected while the queue still has free slots.

With N gossip handlers calling `queuePendingEnvelope` at once on a queue at `capacity-k`, more than the `k` free slots can observe `count > capacity` and bail out, so fewer than `k` messages are stored.

The sibling `executionPayloadBidService` — which #23644 migrates onto this queue — avoids this with `pendingMu` held across `Load` + cap check + `Store`. Absorbing that service into this queue would be a regression unless the guard is added here.

File: cl/phase1/network/services/pending_job_queue.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+		q.count.Add(-1)
+		return false
+	}
+	return true
+}
+
+func (q *pendingJobQueue[K, M]) storeReserved(key K, msg M) {
+	if _, loaded := q.jobs.LoadOrStore(key, &pendingJob[M]{
+		msg:          msg,
+		creationTime: time.Now(),
```
Comment (AskAlexSharov) [inline cl/phase1/network/services/pending_job_queue.go:90]: `storeReserved` always stamps a fresh `creationTime`, so a job re-enqueued under the same key restarts the 30 s expiry.

`processPending` removes the job before calling `process()`, and `process()` runs `ProcessMessage`, which re-queues when its own dependency check fails. `forkGraphDisk.blocks` is pruned, so a block can disappear between `tryProcessPendingEnvelope`'s `GetBlock` and `ProcessMessage`'s `GetBlock`. Each cycle stores a new job with `time.Now()`, so the expiry never fires and the envelope is retried every 100 ms forever — re-running `OnExecutionPayload(ctx, env, false, true)` each time.

Carrying the original `creationTime` forward on re-enqueue (or checking expiry against a first-seen timestamp) would bound this.

File: cl/phase1/network/services/pending_job_queue.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+	}); loaded {
+		q.count.Add(-1)
+	} else {
+		q.cond.L.Lock()
+		q.cond.Signal()
+		q.cond.L.Unlock()
+	}
+}
+
+func (q *pendingJobQueue[K, M]) remove(key K) {
```
Comment (AskAlexSharov) [inline cl/phase1/network/services/pending_job_queue.go:100]: `remove` deletes whatever is under the key and decrements `count` unconditionally — no check that the stored job is the one `processPending` decided on.

The field comment on line 37-39 claims "A non-nil process func runs after the removal, so that a concurrent re-enqueue under the same key is not lost." That only holds for a re-enqueue triggered by `process()` itself. A concurrent gossip enqueue that reaches `LoadOrStore` *before* this `Delete` sees the old job, takes the `loaded` branch, decrements its own reservation and is dropped — then `remove` deletes the old job anyway. Both messages are gone.

This is benign today only because both key types are content-addressed (the dropped message is byte-identical). `executionPayloadBidService.deletePendingBid` does carry the identity check (`current != job` -> return false); folding it in here keeps the guarantee for the migrations in #23644-#23646, and the comment should be narrowed to what the code actually promises.

File: cl/phase1/network/services/execution_payload_service.go [hunk 7]
```diff
@@ -204,104 +202,31 @@ func (s *executionPayloadService) ProcessMessage(ctx context.Context, _ *uint64,
...
 // queuePendingEnvelope adds an envelope to the pending queue for later processing
 func (s *executionPayloadService) queuePendingEnvelope(blockRoot common.Hash, envelope *cltypes.SignedExecutionPayloadEnvelope) {
-	if s.pendingCount.Add(1) > maxPendingEnvelopes {
-		s.pendingCount.Add(-1)
-		return
-	}
-
-	// Compute envelope hash to allow multiple candidates per block
-	envelopeHash, err := envelope.HashSSZ()
+	err := s.pending.enqueueLazy(envelope, func() (pendingEnvelopeKey, error) {
```
Comment (AskAlexSharov) [inline cl/phase1/network/services/execution_payload_service.go:205]: The `buildKey` closure is allocated on every gossip envelope, including the ones the cap rejects.

`enqueueLazy` is too big to inline, so the closure escapes and is heap-allocated at this call site *before* `reserve()` runs. The old code did the cap check first and allocated nothing on the reject path. The cap exists to bound work under a flood, and this puts an allocation back on exactly that path (same for `queuePendingAttestation`).

Passing the parts instead of a closure — e.g. `enqueue(msg)` plus a `keyOf(msg) (K, error)` method value stored on the queue at construction — keeps the laziness without a per-message allocation.

File: cl/phase1/network/services/pending_job_queue.go [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+// pendingJobQueue holds gossip messages whose processing dependencies have not
+// arrived yet and retries them periodically until processed or expired.
+type pendingJobQueue[K comparable, M any] struct {
+	capacity int32
+	expiry   time.Duration
+	tick     time.Duration
+	// tryProcess decides a job's fate: done=false keeps it queued for the next
+	// tick; done=true removes it. A non-nil process func runs after the removal,
+	// so that a concurrent re-enqueue under the same key is not lost.
+	tryProcess func(ctx context.Context, key K, msg M) (process func(), done bool)
```
Comment (AskAlexSharov) [inline cl/phase1/network/services/pending_job_queue.go:40]: `tryProcess` returning a `func()` to run after removal is an awkward contract, and it allocates a closure per processed job (`tryProcessPendingEnvelope` and `tryProcessPendingAttestation` both build one capturing `ctx`/`key`/`msg`).

A decision plus a separate callback expresses the same remove-before-process ordering with no indirection and no allocation:

```go
type jobDecision int // jobKeep, jobDrop, jobReady

tryProcess func(ctx context.Context, key K, msg M) jobDecision
process    func(ctx context.Context, key K, msg M)
```

`processPending` then does `remove(k)` and calls `q.process(ctx, k, job.msg)` on `jobReady`.

File: cl/phase1/network/services/execution_payload_service_test.go [hunk 9]
```diff
@@ -82,7 +82,7 @@ func TestExecutionPayloadServiceBlockNotFound(t *testing.T) {
 
 	// Verify envelope was queued (check internal state)
 	impl := service.(*executionPayloadService)
-	require.Equal(t, int32(1), impl.pendingCount.Load())
+	require.Equal(t, int32(1), impl.pending.count.Load())
 
 	// Now add block to forkchoice
 	fcu.Blocks[blockRoot] = &cltypes.SignedBeaconBlock{
```
Comment (AskAlexSharov) [inline cl/phase1/network/services/execution_payload_service_test.go:88]: This writes `fcu.Blocks` (a plain `map`) while the service's background `loop` goroutine — started by `setupExecutionPayloadService` via `NewExecutionPayloadService(t.Context(), ...)` — reads it through `GetBlock` every 100 ms from `tryProcessPendingEnvelope`.

That is an unsynchronized concurrent map read/write. The window is narrow (I did not hit it in `-race -count=3`), but the `qa-test-race` workflow can, and it would look like an unrelated flake.

Either use a service built without the loop (as the other pending tests in this file now do) or add the block before `ProcessMessage` queues it.

File: cl/phase1/network/services/pending_job_queue.go [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+		capacity:   capacity,
+		expiry:     expiry,
+		tick:       tick,
+		tryProcess: tryProcess,
+		onExpired:  onExpired,
+		cond:       sync.NewCond(&sync.Mutex{}),
+	}
+}
+
+func (q *pendingJobQueue[K, M]) enqueueLazy(msg M, buildKey func() (K, error)) error {
```
Comment (AskAlexSharov) [inline cl/phase1/network/services/pending_job_queue.go:65]: The new shared type has no direct test — its contract is only exercised through the two services.

Untested paths: the `buildKey` error return (neither service test forces `HashSSZ` to fail), `loop`'s wake-on-signal and context-cancel exit, and the duplicate-key `loaded` branch of `storeReserved`. The service tests reach into `pending.jobs` / `pending.count` and call `processPending` directly, so they pin the internals without covering the queue's own API.

A `pending_job_queue_test.go` with a trivial `K=int, M=int` instantiation would cover all of it cheaply and stop the next three migrations from re-deriving the behaviour from the service tests.
